### PR TITLE
MSI-2724: Inventory is not updated when Product SKU is changed.

### PR DESCRIPTION
**Pull request by nmalevanec**
### Fixed Issues (if relevant)

1. magento/inventory#2724: Inventory is not updated when Product SKU is changed

### Manual testing scenarios (*)

1. ...
2. ...

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

File: app/code/Magento/InventoryCatalog/Plugin/Catalog/Model/ProductRepository/DeleteSourceItemsPlugin.php [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+     * @param bool $result
+     * @param ProductInterface $product
+     * @return bool
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     */
+    public function afterDelete(ProductRepositoryInterface $subject, $result, ProductInterface $product): bool
+    {
+        try {
+            $this->deleteSourceItemsBySkus->execute([(string)$product->getSku()]);
+        } catch (CouldNotSaveException|InputException $e) {
```
Comment (sidolov) [inline app/code/Magento/InventoryCatalog/Plugin/Catalog/Model/ProductRepository/DeleteSourceItemsPlugin.php]: Please, use `\Magento\Framework\Exception\CouldNotDeleteException` instead

File: app/code/Magento/InventoryCatalog/Plugin/Catalog/Model/ResourceModel/Product/ProcessSourceItemsPlugin.php [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+     */
+    public function afterSave(
+        Product $subject,
+        Product $result,
+        AbstractModel $product
+    ): Product {
+        if ($this->areSourceItemsShouldBeDeleted($product)) {
+            try {
+                $this->deleteSourceItemsBySkus->execute([(string)$product->getOrigData('sku')]);
+            } catch (CouldNotSaveException|InputException $e) {
```
Comment (sidolov) [inline app/code/Magento/InventoryCatalog/Plugin/Catalog/Model/ResourceModel/Product/ProcessSourceItemsPlugin.php:70]: Please, use `\Magento\Framework\Exception\CouldNotDeleteException` instead

File: app/code/Magento/InventoryCatalog/Model/DeleteSourceItemsBySkus.php [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        DeleteMultiple $sourceItemsDelete
+    ) {
+        $this->getSourceItemsBySku = $getSourceItemsBySku;
+        $this->sourceItemsDelete = $sourceItemsDelete;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function execute(array $skus): void
```
Comment (sidolov) [inline app/code/Magento/InventoryCatalog/Model/DeleteSourceItemsBySkus.php]: For bulk operations, we should use an approach similar to https://github.com/magento/inventory/pull/2728. 
In this case, we should introduce result object and handle errors for each sku in the array.

**Review by sidolov — CHANGES_REQUESTED** [review]

File: app/code/Magento/InventoryCatalog/Plugin/Catalog/Model/ProductRepository/DeleteSourceItemsPlugin.php (hunk 1 above)
Comment (nmalevanec) [reply app/code/Magento/InventoryCatalog/Plugin/Catalog/Model/ProductRepository/DeleteSourceItemsPlugin.php]: @sidolov, after introducing ResultInterface nor relevant anymore.

File: app/code/Magento/InventoryCatalog/Plugin/Catalog/Model/ResourceModel/Product/ProcessSourceItemsPlugin.php (hunk 2 above)
Comment (nmalevanec) [reply app/code/Magento/InventoryCatalog/Plugin/Catalog/Model/ResourceModel/Product/ProcessSourceItemsPlugin.php:70]: @sidolov, after introducing ResultInterface nor relevant anymore.

File: app/code/Magento/InventoryCatalog/Model/DeleteSourceItemsBySkus.php (hunk 3 above)
Comment (nmalevanec) [reply app/code/Magento/InventoryCatalog/Model/DeleteSourceItemsBySkus.php]: @sidolov, done

**Comment by sidolov** [comment]
@magento run Functional Tests EE,Unit Tests

**Comment by sidolov** [comment]
@magetno run Functional Tests EE

**Comment by sidolov** [comment]
@magento run Functional Tests EE

**Review by sidolov — APPROVED** [review]

**Comment by sidolov** [comment]
@magento run Functional Tests B2B

**Comment by nmalevanec** [comment]
see https://github.com/magento/inventory/pull/2795

**Pull request closed without merging**